fs.js

```diff
@@ -0,0 +1,278 @@
+		return new Path(pathModule.dirname(this.path));
+	}
+	read(options = {}) {
+		return new Promise((resolve, reject) => {
+			fs.readFile(this.path, options, (err, data) => {
+				err ? reject(err) : resolve(data);
+			});
+		});
+	}
+	readSync(options = {}) {
+		return fs.readFileSync(this.path, options);
+	}
+	readTextIfExists() {
+		return new Promise((resolve, reject) => {
+			fs.readFile(this.path, 'utf8', (err, data) => {
+				if (err && err.code === 'ENOENT') return resolve('');
+				err ? reject(err) : resolve(data);
+			});
+		});
+	}
+	readTextIfExistsSync() {
+		try {
+			return fs.readFileSync(this.path, 'utf8');
+		} catch (err) {
+			if (err.code !== 'ENOENT') throw err;
+		}
+		return '';
+	}
+	/**
+	 * @param {string | Buffer} data
+	 * @param {Object} options
+	 */
+	write(data, options = {}) {
+		if (Config.nofswriting) return Promise.resolve();
+		return new Promise((resolve, reject) => {
+			fs.writeFile(this.path, data, options, err => {
+				err ? reject(err) : resolve();
+			});
+		});
+	}
+	/**
+	 * @param {string | Buffer} data
+	 * @param {Object} options
+	 */
+	writeSync(data, options = {}) {
+		if (Config.nofswriting) return;
+		return fs.writeFileSync(this.path, data, options);
+	}
+	/**
+	 * @param {string | Buffer} data
+	 * @param {Object} options
+	 */
+	append(data, options = {}) {
+		if (Config.nofswriting) return Promise.resolve();
+		return new Promise((resolve, reject) => {
+			fs.appendFile(this.path, data, options, err => {
+				err ? reject(err) : resolve();
+			});
+		});
+	}
+	/**
+	 * @param {string | Buffer} data
+	 * @param {Object} options
+	 */
+	appendSync(data, options = {}) {
+		if (Config.nofswriting) return;
+		return fs.appendFileSync(this.path, data, options);
+	}
+	/**
+	 * @param {string} target
+	 */
+	symlinkTo(target) {
+		if (Config.nofswriting) return Promise.resolve();
+		return new Promise((resolve, reject) => {
+			// @ts-ignore TypeScript bug
+			fs.symlink(target, this.path, err => {
+				err ? reject(err) : resolve();
+			});
+		});
+	}
+	/**
+	 * @param {string} target
+	 */
+	symlinkToSync(target) {
+		if (Config.nofswriting);
+		return fs.symlinkSync(target, this.path);
+	}
+	/**
+	 * @param {string} target
+	 */
+	rename(target) {
+		if (Config.nofswriting) return Promise.resolve();
+		return new Promise((resolve, reject) => {
+			fs.rename(target, this.path, err => {
+				err ? reject(err) : resolve();
+			});
+		});
+	}
+	/**
+	 * @param {string} target
+	 */
+	renameSync(target) {
+		if (Config.nofswriting) return;
+		return fs.renameSync(target, this.path);
+	}
+	readdir() {
+		return new Promise((resolve, reject) => {
+			fs.readdir(this.path, (err, data) => {
+				err ? reject(err) : resolve(data);
+			});
+		});
+	}
+	readdirSync() {
+		return fs.readdirSync(this.path);
+	}
+	createWriteStream(options = {}) {
+		if (Config.nofswriting) {
+			const Writable = require('stream').Writable;
+			return new Writable({write: (chunk, encoding, callback) => callback()});
```

Comment:
The default ``WriteStream`` from ``fs`` *is* async (what else?), without need of any ``setImmediate`` call.
Using a boilerplate above it which is sometimes sync and sometimes async is a well-known antipattern with all sorts of bad effects regarding timings when using any such API.

http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony